Issue 2846 ie8 failures

shane-tomlinson opened the pull request:
@ozten - can you review this?

Fix IE8 errors
- Added a new test helpers - testElementTextContains.
- Use the new helper instead of text.indexOf to check for text in the DOM.
- Remove extra variables that were not used.
- Fix the thrown exception when focusing the password field in the unit tests.
- Remove the console.log message for `actions called={"doAuthenticate":true}`

fixes #2846
Partial fix for #2828

Commit:
Fix IE8 not finding the correct text when a primary goes offline.

* Added a new test helpers - testElementTextContains.
* Use this in the test for primary_offline to see if the primary name has been written to the DOM.
* Remove extra variables that were not used.

Commit:
Fix IE8 test failures

* console.log does not exist in IE8
* Use testElementTextContains to check whether an element has text.

Commit:
Fix IE8 blowing up when trying to focus the password field during unit tests.

shane-tomlinson commented on resources/static/test/cases/dialog/js/modules/set_password.js:
```diff
@@ -63,9 +64,9 @@
       transition_no_password: true
     });
-    var msg = $("#set_password .inputs li").text();
-    ok(msg.indexOf("no longer allows"), "transition message shown");
```
String.prototype.indexOf returns -1 if a value is not found. The index for this text in IE8 was 0.

ozten commented on resources/static/test/cases/dialog/js/misc/state.js:
```diff
@@ -480,7 +480,6 @@
       type: 'secondary',
       state: 'transition_to_secondary'
     });
-    console.log('actions called=' + JSON.stringify(actions.called));
```
Doh. Thanks.

ozten commented:
I'm r+, but you might want to do the following to the test suite:

1) In the testing framework, detect that we're using ie8 by simulating this failure and setting a `ie8` flag.

2) For all browsers, run through the steps, to catch dom exceptions, but

3) For tests which ie8 will get an empty string, skip checking for the expected string. Alternatively have a check for empty string, you're call.

Commit:
Two changes due to @ozten's suggestions.

1) Make sure testElementTextContains does not blow up if the element is not found
2) Dig deeper into the ie8 focus exception to realize the exception was thrown because the DOM was no longer available. If we remove the timing problem, the exception goes away and the try/catch is no longer needed.

shane-tomlinson commented:
@ozten - Thanks for the review, I made two changes from your suggestions.

1) in testElementTextContains and testElementTextEquals, make sure there are no explosions if the elements to get the text from are not found.

2) Look more closely at the cause of the focus exception in IE8. The cause was due to a timing error where complete was called before the focus occurred. When complete called, the DOM got torn down before focus occurred. Since the element was no longer in the DOM, focus excepted. Since I moved complete into the showHint callback already, the problem was already side stepped and the try/catch was unnecessary. I removed it. This means no IE8 specific testing for the exception.

ozten commented:
Thanks for jumping on 2852, great work!
